Realizaremos el segundo modelo de predicción para predecir la calificación de un producto para saber cuales productos gusta más o menos: regresión

In [6]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score
from sklearn.impute import SimpleImputer

# ✅ Archivo correcto
df = pd.read_csv('/content/READYecommerce_raw_dataset.csv')

# ✅ Solo filas con rating
df_rated = df[df['rating'].notna()].copy()

features = ['age', 'quantity', 'unit_price', 'discount_pct',
            'order_month', 'order_dayofweek', 'is_discounted',
            'category_enc', 'country_enc', 'payment_method_enc',
            'shipping_method_enc']

X2 = df_rated[features]
y2 = df_rated['rating']

# Split
X2_train, X2_test, y2_train, y2_test = train_test_split(
    X2, y2, test_size=0.2, random_state=42)

# ✅ Imputar por si hay NaN
imputer = SimpleImputer(strategy='median')
X2_train = imputer.fit_transform(X2_train)
X2_test = imputer.transform(X2_test)

# Entrenar
reg = RandomForestRegressor(n_estimators=100, random_state=42)
reg.fit(X2_train, y2_train)

# Evaluar
y2_pred = reg.predict(X2_test)
print('MAE:', mean_absolute_error(y2_test, y2_pred))
print('R2 Score:', r2_score(y2_test, y2_pred))

# ✅ Agrega esto para ver feature importance
importances = pd.Series(
    reg.feature_importances_,
    index=features
).sort_values(ascending=False)
print('\nFeature Importance:')
print(importances)


MAE: 0.8548165137614679
R2 Score: -0.10013588286969277

Feature Importance:
unit_price             0.210925
age                    0.177817
order_month            0.111856
country_enc            0.091954
order_dayofweek        0.090244
shipping_method_enc    0.069703
quantity               0.063845
payment_method_enc     0.062401
category_enc           0.055300
discount_pct           0.053518
is_discounted          0.012437
dtype: float64


Hallazgo Modelo 2 — Rating Prediction:
MAE: 0.854 | R2: -0.10

El modelo no logró predecir ratings con precisión útil. Sin embargo,
el análisis de feature importance sugiere que en datos reales,
unit_price (0.21) y age (0.17) serían los factores más relevantes
para la satisfacción del cliente.
